src/charm.py

```diff
@@ -1,117 +1,278 @@
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            self.unit.status = BlockedStatus("Waiting for TLS.")
+            return
+        self.unit.status = ActiveStatus()
+
+    def _on_config_changed(self, _: Any) -> None:
+        """Handle the config-changed event."""
+        self._reconcile()
+
+    def _on_start(self, _: Any) -> None:
+        """Handle the start event."""
+        self._reconcile()
+
+    def _on_get_certificate_action(self, event: ActionEvent) -> None:
+        """Triggered when users run the `get-certificate` Juju action.
+
+        Args:
+            event: Juju event
+        """
+        hostname = event.params["hostname"]
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            event.fail("Certificates relation not created.")
+            return
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
-        """Define and start a workload using the Pebble API.
+        tls_rel_data = tls_certificates_relation.data[self.app]
+        if not tls_rel_data.get(f"certificate-{hostname}"):
+            event.fail("Certificate not available")
+            return
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+        event.set_results(
+            {
+                f"certificate-{hostname}": tls_rel_data[f"certificate-{hostname}"],
+                f"ca-{hostname}": tls_rel_data[f"ca-{hostname}"],
+                f"chain-{hostname}": tls_rel_data[f"chain-{hostname}"],
+            }
+        )
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+    def _on_certificates_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the TLS Certificate relation created event.
 
         Args:
-            event: event triggering the handler.
+            event: The event that fires this method.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
+        if not self._are_relations_ready():
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        hostname = self.get_hostname()
+        if not hostname:
+            self.unit.status = BlockedStatus("Waiting for hostname to be configured")
+            event.defer()
```

Comment:
I saw Arturo's comment and I believe in this case the defer is not needed, right? Since a config changed will be received after setting hostname.

Reply:
I keep `defer` here because I think this handler should only run once. Here it's creating the juju secret that will be used to create the CSR. If we remove `defer` here then we'd need to move also the event handler code into the main `reconcile` loop. which might make it a bit too complex with the k8s resource management logic that we'd also want to put there.

What do you think? should we keep it like this or should we remove `defer` and move the below code into the main `reconcile` loop?

Reply:
Hm...I see your point.
Well, if in the future you can guarantee that the hostname will always be set (that charm state + validator comment somewhere), then this could be removed right?

Reply:
yes, I already have a PR (a bit down the line since it's like 4 PRs pending before it) to refactor the charm state logic and add pydantic validation so that it can be removed.